### PR TITLE
EV-21 Added Ability table to the question model and updated relations

### PR DESCRIPTION
**Pull request by JackHKJ**
## What is the current behavior?

Scalar representation of the difficulty of solution / ability of user

Issue Number: EV-21

User 1:1 Ability (as ability in the User table)
Solution 1:1 Ability (as difficulty in the Solution table)

## Pull request type

DB Schema Update

### CODE REVIEW

File: src/backend/EX/settings.py [hunk 1]
```diff
@@ -186,3 +188,83 @@
 # Other
 LOGIN_REDIRECT_URL = 'home'
 ACCOUNT_LOGOUT_REDIRECT_URL = 'home'
+
+# Ability Constants
+VALID_ABILITY_KEYS = {
```
Comment (Windsooon) [inline src/backend/EX/settings.py:193]: If the key in the HashTable always equals the value, we can use HashSet instead.

File: src/backend/common/models.py [hunk 2]
```diff
@@ -4,14 +4,27 @@
 from django.conf import settings
 from question.models import Solution
 
+
 class User(AbstractUser):
-    avatar = models.URLField(max_length=1024, blank=True, help_text="URL for avatar")
+    avatar = models.URLField(max_length=1024, blank=True,
+                             help_text="URL for avatar")
+    abilityId = models.ForeignKey(
```
Comment (Windsooon) [inline src/backend/common/models.py:11]: According to the naming convention of PEP8, use `ability_id`, `user_id`, and `question_id` instead.

File: src/backend/common/models.py [hunk 3]
```diff
@@ -4,14 +4,27 @@
...
 class UserSubmission(models.Model):
-    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE, related_name='submission_user')
-    solution = models.ForeignKey(Solution, on_delete=models.CASCADE, related_name='submission_solution')
-    content = models.JSONField()
+    # Representation of the userId (settings.AUTH_USER_MODEL.id)
+    userId = models.BigIntegerField()
+    # Representation of the question id (Question.id)
+    questionId = models.BigIntegerField()
+    # List of category ids, should be validated
+    categoryIds = models.JSONField()
```
Comment (Windsooon) [inline src/backend/common/models.py:21]: For simplicity and consistency, we use the singular form, so it should be `category_id`

File: src/backend/question/admin.py [hunk 4]
```diff
@@ -1,21 +1,25 @@
...
+from .models import Question, Keypoint, UserKeypointScore, Solution, Category
+
 
 class QuestionAdmin(admin.ModelAdmin):
     list_display = ('id', 'name', 'type')
 
+
 class KeypointAdmin(admin.ModelAdmin):
-    list_display = ('id', 'name', 'category')
+    list_display = ('id', 'name', 'categoryId')
```
Comment (Windsooon) [inline src/backend/question/admin.py:10]: use `category_id` instead

File: src/backend/question/models.py [hunk 5]
```diff
@@ -11,62 +13,98 @@ class Category(models.Model):
...
         SHORT_ANSWER = 'SA', 'Short Answer'
         TRUE_O_FALSE = 'TF', 'True / False'
 
-    name = models.CharField(max_length=100, blank=True, help_text='question name')
+    name = models.CharField(max_length=100, blank=True,
+                            help_text='question name')
     description = models.JSONField(help_text='description of question')
-    category = models.ManyToManyField(Category)
-    type = models.CharField(max_length=2, choices=QuestionTypeChoice.choices, help_text='type of question')
+    categoryIds = models.ManyToManyField(Category)
```
Comment (Windsooon) [inline src/backend/question/models.py:27]: also change to `category_id`

File: src/backend/question/models.py [hunk 6]
```diff
@@ -11,62 +13,98 @@ class Category(models.Model):
...
+    type = models.CharField(
```
Comment (Windsooon) [inline src/backend/question/models.py:28]: Does the `Question` have a type?  type should only used in `Solution`

File: src/backend/question/models.py [hunk 7]
```diff
@@ -11,62 +13,98 @@ class Category(models.Model):
...
 
     def __str__(self):
         return f'{self.name}'
 
+
+class Ability(models.Model):
+    ability: models.JSONField(default=dict, blank=True)
+
+    # Sanity check on save
+    def clean(self):
```
Comment (Windsooon) [inline src/backend/question/models.py:63]: Why we need `clean` and `save` function here?

**Review by Windsooon — CHANGES_REQUESTED** [review]

**Comment by Windsooon** [comment]
@JackHKJ we can rebase the commit into two commits:
1. files changed related to admin.py and settings.py
2. files changed related to model.py and migrations.py

File: src/backend/EX/settings.py (hunk 1 above)
Comment (JackHKJ) [reply src/backend/EX/settings.py:193]: Using a dictionary allows u utilize the hint in IDE for better type support, rather hash set needs additional step for retrieving a key.

**Comment by JackHKJ** [comment]
This PR is closed due to inappropriate granuality

**Pull request closed without merging**